### PR TITLE
Move yamlRestCompatTest plugin to internal and add a new lifecycle

### PR DESCRIPTION
**Pull request by jakelandis**
To prepare for the introduction of REST API compatibility testing, this commit exposes the YamlRestCompatTestPlugin for use and ties it to a new lifecycle task `checkRestCompat`. Since REST API compatibility will double the time it takes to run all REST tests, it will be run outside of the normal 'check' task.  This new lifecycle will need to be mapped to a CI job run in parallel to the normal 'check' task. Currently only `yamlRestCompatTest` is mapped to this lifecycle, but future version may require additional tasks and the abstraction can be used without the need to change CI.

### CODE REVIEW

**Review by pgomulka — APPROVED** [review]
LGTM
Only thing to consider. 
What if we include unit tests from `:x-pack:rest-compatibility` (the tests for V7/N-1 code) under this check as well (possibly a rename to `checkCompat`? )

**Comment by mark-vieira** [comment]
How BWC testing works now is that `check` is wired up to run only snapshot BWC versions. I think we want to continue this as most of our periodic builds just run `check` and we'll continue to have that be the case. For intake and PRs we run `checkPart1` and `checkPart2` which run OSS and Xpack, respectively. We then separately run `bwcTestSnapshots`. For the `checkPart*` tasks, when we run them we actually[ explicitly _disable_ BWC testing](https://github.com/elastic/elasticsearch/blob/master/build.gradle#L204-L207) so that `check` in that case doesn't run BWC. This is kinda nasty, but let's follow the same pattern here:

1. Running `check` by default runs the compatibility tests on snapshot versions.
2. Running `checkPart*` disables (1)
3. Add a separate CI job to intake/PRs that runs just the rest compat tests.

@jakelandis forgive my confusion here, it's been awhile since I've thought about this but what all BWC versions are we handling here? Do we create per-version test tasks like we do for other BWC stuff? For example, the way I understand it, is for `master` we should be running the _tests_ from compatible branches against `master` code. Which branches will that be?

**Comment by mark-vieira** [comment]
> +1 , thanks for the guidance. Do think I should keep the 'checkRestCompat' lifecycle task as an abstraction in case we decide to add other things beyond yamlRestCompatTest task under the rest compatibility testing umbrella ?

Yeah, I don't see why not. We can easily remove it if it becomes annoying.

> However, I am having doubts that we actually need/want that full suite and probably worth a discussion (in person).

Yeah, this is the never ending debate on more coverage vs efficiency. I'm generally in favor of your approach of starting small and expanding coverage where we discover gaps.

**Comment by mark-vieira** [comment]
I'll also clarify that I think for simplicity we should just have the rest compat tests piggyback on the existing flag for disabling bwc tests. That is, if we set `bwc_tests_enabled=false` this _also_ turns off the rest compatibility tests. If we feel the need to make a distinction here we can introduce a separate flag down the road.

**Comment by jakelandis** [comment]
@mark-vieira @breskeby - This should be ready for review. I kept the lifecycle task, and the enabled status of that task is married to the `bwc_tests_enabled` flag and then I married the `YamlRestCompatTestPlugin` tasks enabled to the enabled state of the new lifecycle task.  A tests is now also included.  (more tests are on https://github.com/elastic/elasticsearch/pull/64186)

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/CheckRestCompatPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/CheckRestCompatPlugin.java:20]: I think this should go in the `internal` package. We don't intent for external folks to use this.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java:71]: No please let's not do this. This opens up the door for all kinds of issues due to configuration ordering. We should skip these tasks at _execution_ time via an `onlyIf` condition. All tasks that do work (the tests) should simply have an `onlyIf` condition on them [as we do with all other bwc tests](https://github.com/elastic/elasticsearch/blob/90e10615af16f2c0cb182938496d24d500e283f4/gradle/bwc-test.gradle#L24).

There's no need to disable/skip the lifecyle tasks as the don't actually do work.

**Review by mark-vieira — COMMENTED** [review]
Added a couple of comments.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/CheckRestCompatPlugin.java
Comment (jakelandis) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/CheckRestCompatPlugin.java]: moved: https://github.com/elastic/elasticsearch/pull/67394/commits/322b0b1197c8df95368ea9b99c125d938f9d4beb

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java
Comment (jakelandis) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java]: used onlyIf instead of enabled=false, removed the disabling of the lifecycle task, and allow the YamlRestCompatTestPlugin to read the ext directly. https://github.com/elastic/elasticsearch/pull/67394/commits/176a72ed820d97948ab82e7bd736bf9201ef6be6

**Comment by jakelandis** [comment]
thanks @mark-vieira - the changes have been implemented and is ready for another review.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java:20]: I missed a spot. I think this should be `internal` as well.

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/CheckRestCompatPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/internal/CheckRestCompatPlugin.java:32]: @jakelandis any strong motivation for this being a separate plugin rather than just folding this into `YamlRestCompatTestPlugin`. I have no issue with plugins like that _also_ implementing lifecycle conventions.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java:163]: @breskeby is this method of accessing extra properties kosher with the configuration cache?

**Review by mark-vieira — APPROVED** [review]
Couple of comments, otherwise LGTM.

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/CheckRestCompatPlugin.java
Comment (jakelandis) [reply buildSrc/src/main/java/org/elasticsearch/gradle/internal/CheckRestCompatPlugin.java:32]: nope ... no strong motivation just thought lifecycle plugins would be more standalone entities. I can move the task over.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java:163]: I think what breaks configuration cache here would be accessing project. I think we need to change how `bwc_tests_enabled` is accessed in general and not just here. So I'm fine to keep this for now here. We will revisit when we rework the general way of accessing build properties in a configuration cache friendly manner

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java
Comment (jakelandis) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java]: done on https://github.com/elastic/elasticsearch/pull/67394/commits/7c5be906d0306af47e0646c49a835215cb342d14 , had to add getters/setters since alot of access was via package private directly.

**Comment by jakelandis** [comment]
thanks @mark-vieira and @breskeby - comments addressed and will merge once it turns green.

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/CheckRestCompatPlugin.java
Comment (jakelandis) [reply buildSrc/src/main/java/org/elasticsearch/gradle/internal/CheckRestCompatPlugin.java]: done on https://github.com/elastic/elasticsearch/pull/67394/commits/81b333076f8d6a3ee98d9ba01740995b0a8ad8a6

**Comment by jakelandis** [comment]
@elasticmachine run elasticsearch-ci/2  (offending test has been muted in other PR)

**Comment by jakelandis** [comment]
@elasticmachine update branch

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java:193]: Agreed. I think that should be a property on `BuildParams` instead.

Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java:193]: Agreed. We should just move it to `gradle.properties` really instead of using extra properties for this.

**Pull request merged**
